node/bindnode/node.go

```diff
@@ -0,0 +1,1112 @@
+		// fmt.Println(kval.Interface(), val.Interface())
+		w.val.FieldByName("Index").SetInt(int64(idx))
+		w.val.FieldByName("Value").Set(val)
+		return nil
+	}
+	return &_assembler{
+		schemaType: mtyp,
+		val:        val,
+		finish:     finish,
+	}
+}
+
+func (w *_unionAssembler) AssembleEntry(k string) (ipld.NodeAssembler, error) {
+	if err := w.AssembleKey().AssignString(k); err != nil {
+		return nil, err
+	}
+	am := w.AssembleValue()
+	return am, nil
+}
+
+func (w *_unionAssembler) Finish() error {
+	if w.finish != nil {
+		if err := w.finish(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (w *_unionAssembler) KeyPrototype() ipld.NodePrototype {
+	return &_prototype{schemaType: schemaTypeFieldName, goType: goTypeString}
+}
+
+func (w *_unionAssembler) ValuePrototype(k string) ipld.NodePrototype {
+	panic("TODO: struct ValuePrototype")
+}
+
+type _structIterator struct {
+	// TODO: support embedded fields?
+	schemaType *schema.TypeStruct
+	fields     []schema.StructField
+	val        reflect.Value // non-pointer
+	nextIndex  int
+
+	// these are only used in repr.go
+	reprEnd int
+}
+
+func (w *_structIterator) Next() (key, value ipld.Node, _ error) {
+	if w.Done() {
+		return nil, nil, ipld.ErrIteratorOverread{}
+	}
+	field := w.fields[w.nextIndex]
+	val := w.val.Field(w.nextIndex)
+	w.nextIndex++
+	key = basicnode.NewString(field.Name())
+	if field.IsOptional() {
+		if val.IsNil() {
+			return key, ipld.Absent, nil
+		}
+		val = val.Elem()
+	}
+	if field.IsNullable() {
+		if val.IsNil() {
+			return key, ipld.Null, nil
+		}
+		val = val.Elem()
+	}
+	node := &_node{
+		schemaType: field.Type(),
+		val:        val,
+	}
+	return key, node, nil
+}
+
+func (w *_structIterator) Done() bool {
+	return w.nextIndex >= len(w.fields)
+}
+
+type _listIterator struct {
+	schemaType *schema.TypeList
+	val        reflect.Value // non-pointer
+	nextIndex  int
+}
+
+func (w *_listIterator) Next() (index int64, value ipld.Node, _ error) {
+	if w.Done() {
+		return 0, nil, ipld.ErrIteratorOverread{}
+	}
+	idx := int64(w.nextIndex)
+	val := w.val.Index(w.nextIndex)
+	w.nextIndex++
+	return idx, &_node{schemaType: w.schemaType.ValueType(), val: val}, nil
+}
+
+func (w *_listIterator) Done() bool {
+	return w.nextIndex >= w.val.Len()
+}
+
+type _unionIterator struct {
+	// TODO: support embedded fields?
+	schemaType *schema.TypeUnion
+	members    []schema.Type
+	val        reflect.Value // non-pointer
+
+	done bool
+}
+
+func (w *_unionIterator) Next() (key, value ipld.Node, _ error) {
+	if w.Done() {
+		return nil, nil, ipld.ErrIteratorOverread{}
+	}
+	w.done = true
+
+	haveIdx := int(w.val.FieldByName("Index").Int())
+	mtyp := w.members[haveIdx]
+	mval := w.val.FieldByName("Value").Elem()
+
+	node := &_node{
+		schemaType: mtyp,
+		val:        mval,
+	}
+	key = basicnode.NewString(mtyp.Name().String())
+	return key, node, nil
+}
+
+func (w *_unionIterator) Done() bool {
+	return w.done
+}
+
+// TODO: consider making our own Node interface, like:
```

Comment:
given the length of this file, it might be worth splitting it into per-type files. e.g `node_union`, `node_map`, `node_list` for the type-specific assemblers / iterators.

Reply:
Splitting is a good point. I just kept going and didn't really think about file sizes that much.

Maybe node_scalar and node_recursive would be better for now, and later on I can split recursive into map/list/union if necessary. It feels a bit silly to have e.g. node_bool or node_int.